downsampler/src/main/scala/filodb.downsampler/PartitionDownsampler.scala

```diff
@@ -0,0 +1,106 @@
+  def downsample(shardStats: TimeSeriesShardStats,
+                 resolutions: Seq[FiniteDuration],
+                 downsampleSchema: Schema,
+                 downsamplers: Seq[ChunkDownsampler],
+                 partToDownsample: ReadablePartition,
+                 blockHolder: BlockMemFactory,
+                 bufferPool: WriteBufferPool,
+                 nativeMemManager: MemFactory,
+                 downsampleIngestionTime: Long,
+                 userStartTime: Long,
+                 userEndTime: Long): Map[FiniteDuration, TimeSeriesPartition] = {
+
+    val resToPartition = resolutions.map { res =>
+      val part = new TimeSeriesPartition(
+        0,
+        downsampleSchema,
+        partToDownsample.partitionKey,
+        0,
+        bufferPool,
+        shardStats,
+        nativeMemManager,
+        1)
+      logger.trace(s"Creating new part=${part.hashCode}")
+      res -> part
+    }.toMap
+
+    val chunksets = partToDownsample.infos(AllChunkScan)
+    val timestampCol = 0
+    // TODO create a rowReader that will not box the vals below
+    val downsampleRow = new Array[Any](downsamplers.size)
+    val downsampleRowReader = SeqRowReader(downsampleRow)
+
+    while (chunksets.hasNext) {
+      val chunkset = chunksets.nextInfo
+      val startTime = chunkset.startTime
+      val endTime = chunkset.endTime
+      val vecPtr = chunkset.vectorPtr(timestampCol)
+      val tsReader = partToDownsample.chunkReader(timestampCol, vecPtr).asLongReader
+
+      // for each downsample resolution
+      resToPartition.foreach { case (resolution, part) =>
+        val resMillis = resolution.toMillis
+        var pStart = ((startTime - 1) / resMillis) * resMillis + 1 // inclusive startTime for downsample period
+        var pEnd = pStart + resMillis // end is inclusive
+        // for each downsample period
+        while (pStart <= endTime) {
+          if (pEnd >= userStartTime && pEnd <= userEndTime) {
+            // fix the boundary row numbers for the downsample period by looking up the timestamp column
+            val startRowNum = tsReader.binarySearch(vecPtr, pStart) & 0x7fffffff
+            val endRowNum = Math.min(tsReader.ceilingIndex(vecPtr, pEnd), chunkset.numRows - 1)
+
+            // for each downsampler, add downsample column value
+            for {col <- downsamplers.indices optimized} {
+              val downsampler = downsamplers(col)
+              downsampler match {
+                case d: TimeChunkDownsampler =>
+                  downsampleRow(col) = d.downsampleChunk(partToDownsample, chunkset, startRowNum, endRowNum)
+                case d: DoubleChunkDownsampler =>
+                  downsampleRow(col) = d.downsampleChunk(partToDownsample, chunkset, startRowNum, endRowNum)
+                case h: HistChunkDownsampler =>
+                  downsampleRow(col) = h.downsampleChunk(partToDownsample, chunkset, startRowNum, endRowNum)
+                    .serialize()
```

Comment:
Why does only the histogram downsampler have `serialize()`, what does this do?

Reply:
Oh never mind.  I'm not sure if this will work correctly, as the histogram ingestion is specifically tested and designed to work from BinaryRecords mostly.

Reply:
I just brought this over, as is, from the old downsampler. I believe you wrote it. Does it need a fix ?